File: src/Shape.ts
import Point from './Point';

interface Shape {
  width: number;
  height: number;
  offset(): Point;
  occupiedArea(): number[][];
}

export default Shape;

export const pointsList = (shape: Shape): Array<[Point, number]> => {
  const points: Array<[Point, number]> = [];

  shape.occupiedArea().forEach((row, y) =>
    row.forEach((value, x) => {
      points.push([new Point(x, y).add(shape.offset()), value]);
    }));

  return points;
};

export const filteredPoints
  = (fn: (value: number) => boolean, shape: Shape): Point[] => {
  return pointsList(shape)
    .filter((entry) => fn(entry[1]))
    .map((entry) => entry[0]);
};

export const occupiedPoints = (shape: Shape): Point[] => {
  return filteredPoints((value) => value > 0, shape);
};

export const unoccupiedPoints = (shape: Shape): Point[] => {
  return filteredPoints((value) => value === 0, shape);
};

File: src/Shape.py
from abc import ABCMeta, abstractmethod

from .Point import Point

class Shape(metaclass=ABCMeta):
  def __init__(self, width, height):
    self.width = width
    self.height = height

  def pointsList(self):
    points = []

    for y, row in enumerate(self.occupiedArea()):
      for x, value in enumerate(row):
        points.append((Point(x, y).add(self.offset()), value))

    return points

  def filteredPoints(self, fn):
    return list(entry[0] for entry in self.pointsList() if fn(entry[1]))

  def occupiedPoints(self):
    return self.filteredPoints(lambda value: value > 0)

  def unoccupiedPoints(self):
    return self.filteredPoints(lambda value: value == 0)

  @abstractmethod
  def offset(self):
    pass

  @abstractmethod
  def occupiedArea(self):
    pass

  def __str__(self):
    return '\n'.join(''.join(str(v) for v in row) for row in self.occupiedArea())

  def __repr__(self):
    return str(self)